### ISSUE TITLE
Add a version of operator[] to ea::hash_map that returns a pointer.

### ISSUE THREAD

**Issue by JTippetts** [rokups/rbfx#311]
The vanilla implementation of HashMap implements an operator[] that returns a pointer, or nullptr if the key is not found, which works on a const object. See https://github.com/urho3d/Urho3D/blob/master/Source/Urho3D/Container/HashMap.h#L343

This makes it easier when, eg, parsing JSON, as you can use the construct:

```
JSONValue *thingy=obj["Thingy"];
if(thingy)
{
  // do stuff with thingy
}
```
The existing [] operators for ea::hash_map don't work in this case, since they don't respect const-ness. Instead, you have to do like

```
auto iter=obj.find("Thingy");
if(iter!=obj.end())
{
  const JSONValue &thingy=iter->second;
  // do stuff with thingy
}
```

which, while minor, is still a small bit annoying.

**Comment by iSLC** [comment]
Iirc hash maps create the element if doesn't exist. That's how STL does it and how i think it should be done. If you change this behavior, you are likely to break a lot of code.

There's usually a `try...` method that yields the null if doesn't exist. To avoid an exception or creating the element.

**Comment by eugeneko** [comment]
Well, it shouldn't be "operator []" because it's really confusing, but we can have method for this.

**Comment by JTippetts** [comment]
Implementing it as a [] operator would help in migration from vanilla Urho,. Right now I'm looking at hundreds of cases of the operator[] usage in a project I'm migrating, since the same behavior affects const VariantMaps as well. Vanilla HashMap implements both operators: one that creates the element, returning a non-const reference, and one that returns a pointer and respects const.  But I guess not breaking existing code is probably more important than easing migration.

**Comment by eugeneko** [comment]
> Implementing it as a [] operator would help in migration from vanilla Urho

Porting a _design mistake_ from `Urho3D::HashMap` to `ea::hash_map` will make porting easier, but I find it hard to justify.

**Comment by JTippetts** [comment]
Is it really a design mistake, though? It doesn't interfere with existing operator[] (they're both non-const), it is still semantically accurate (it's performing an element lookup), and it facilitates cleaner code, which I suspect is why Urho implemented it in their custom container to start with.

**Comment by eugeneko** [comment]
> it is still semantically accurate (it's performing an element lookup)

I would say it's not. Semantics of `[]` assume guaranteed fetching of the element, you can see it all across C and C++.
Moreover, it doesn't match semantics of non-const operator `[]`. Basically two completely different operators that are hidden behing single name for whatever reason.

**Comment by JTippetts** [comment]
Well, without it it's still the case (all throughout C++) that there are 2 different operators, only in the current case one of the operators ([] on a const object) is a compiler error. Still, I understand the reasoning here, and accept it. Given the sheer number of custom shaders I'd have to port also, maybe this project I'm porting is best left alone anyway.

**Comment by eugeneko** [comment]
> Given the sheer number of custom shaders I'd have to port also, maybe this project I'm porting is best left alone anyway.
Yep, I think that having non-const [] won't save any substantial time for migration, given how many changes are there.
I still think we can use a function for this tho.
